File: 139-word-break/word-break.py
class Solution:
    def wordBreak(self, s: str, wordDict: List[str]) -> bool:
        
        n = len(s)
        dp = [False] * (n+1)
        dp[0] = True

        for r in range(1, n+1):
            for j in range(r):
                if dp[j] and s[j:r] in wordDict:
                    dp[r] = True
                    break

        return dp[n]

File: 139-word-break/word-break.java
class Solution {
    public boolean wordBreak(String s, List<String> wordDict) {
        
        int l = s.length();
        boolean[] dp = new boolean[l + 1];

        dp[0] = true;

        Set<String> wordSet = new HashSet<>(wordDict);

        for (int i = 1; i < s.length() + 1; i++ ){
            for (int j = 0; j < i; j++ ){
                if (dp[j] && wordSet.contains(s.substring(j, i))){
                    dp[i] = true;
                    break;
                }
            }
        }
        return dp[l];
    }
}